# Auteurs

- MARTHELY Samuel
- DEBA WANDJI Franck

# Installation des packages et chargement des données

In [ ]:
!wget https://people.irisa.fr/Guillaume.Gravier/teaching/ENSAI/data/francetvinfo.json
!wget https://people.irisa.fr/Guillaume.Gravier/teaching/ENSAI/data/tvinfo-sources.json
!pip instalµl feedparser
!pip install newspaper3k
!python -m spacy download fr_core_news_md

--2025-01-02 10:25:09--  https://people.irisa.fr/Guillaume.Gravier/teaching/ENSAI/data/francetvinfo.json
Resolving people.irisa.fr (people.irisa.fr)... 131.254.254.107
Connecting to people.irisa.fr (people.irisa.fr)|131.254.254.107|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5887584 (5.6M) [application/json]
Saving to: ‘francetvinfo.json’

francetvinfo.json   100%[===================>]   5.61M  5.28MB/s    in 1.1s    

2025-01-02 10:25:11 (5.28 MB/s) - ‘francetvinfo.json’ saved [5887584/5887584]

--2025-01-02 10:25:11--  https://people.irisa.fr/Guillaume.Gravier/teaching/ENSAI/data/tvinfo-sources.json
Resolving people.irisa.fr (people.irisa.fr)... 131.254.254.107
Connecting to people.irisa.fr (people.irisa.fr)|131.254.254.107|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 179 [application/json]
Saving to: ‘tvinfo-sources.json’

tvinfo-sources.json 100%[===================>]     179  --.-KB/s    in 0s      

2025-01-02 10:25:12 (8

# Question 1: Récupération de flux RSS

## Lecture de la base de donnée existante

Ici nous chargeons la base de données des article (francetvinfo) ainsi que le la liste des flux RSS (tvinfo-sources). Nous affichons également la taille de chacun de ces dictionnaires.

- france_tv_data est un dictionnaire comportant 2097 articles dont chacun a pour clé son lien url.

- tv_info_source_data est un dictionnaire à trois clé associées aux catégories (france, europe, économie).
leurs valeurs correspondent aux flux rss versles catégories d'articles considérés.

In [ ]:
import json
from pprint import pprint

# Charger les données JSON
with open('francetvinfo.json', 'r') as f:
    france_tv_data = json.load(f)

# Afficher un aperçu des données
#print(rance_tv_data[:5])  # Affiche les 5 premières entrées



# Charger les données JSON
with open('tvinfo-sources.json', 'r') as g:
    tv_info_source_data = json.load(g)

# Afficher un aperçu des données
#print(test_data[:5])  # Affiche les 5 premières entrées

print(len(france_tv_data))
print(len(tv_info_source_data))

2097
3


In [ ]:
# Flux RSS
pprint(tv_info_source_data)

{'economie': 'https://www.francetvinfo.fr/economie/entreprises.rss',
 'europe': 'https://www.francetvinfo.fr/monde/europe.rss',
 'france': 'https://www.francetvinfo.fr/france.rss'}


Nous affichons ci-dessous deux articles du fichier france_tv_data. Cette exploration nous permet notamment d'identifier comment sont structurés les articles dans le fichier.
Pour chaque article on a la structure suivant:

- 'author' : auteur de l'article,
- 'category': la catégorie de l'article (france, europe, economie)
- 'content' : Le contenue de l'article,
- 'date': date de publication de l'article,
- 'image_link': lien vers l'image d'illustration utilisé dans l'article,
- 'title': Titre de l'article

A noter que la clé de chaque article dans le dictionnaire est donnée par son adresse url.

In [ ]:
# Exemples d'articles présents dans france_tv_data
for lien, article in list(france_tv_data.items())[:2]:
    print(f"Lien : {lien}")
    pprint(article)
    print("=" * 50)  # Séparateur pour plus de lisibilité

Lien : https://www.francetvinfo.fr/sante/prevention/eure-quand-les-elus-apprennent-a-se-defendre_3791635.html#xtor=RSS-3-[france]
{'author': [],
 'category': 'france',
 'content': 'Beaucoup de maires et de secrétaires de mairie ont rencontré au '
            'cours de leurs mandats des problèmes de violence, essentiellement '
            "verbale. C'est le cas de Claire. Un jour, un homme est entré dans "
            "son bureau et a frappé très fort sur son bureau pour l'intimider. "
            '"J\'ai réagi à l\'instinct, j\'ai essayé de discuter, mais ça '
            "s'est envenimé. Heureusement qu'une dame est arrivée à la "
            'permanence et a menacé d\'appeler la gendarmerie", raconte Claire '
            'Carrère-Godebout, maire (SE) de Graveron-Sémerville (Eure).\n'
            '\n'
            'Des professeurs de karaté et un psychologue\n'
            '\n'
            'La plupart de ces élus ne savent pas comment se comporter face à '
            "une situation co

## Acquisition des données par écoute d'un flux RSS

In [ ]:
# (a) se premunir contre le bloquage de commande
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

# (b) recuperer le fichier RSS
import feedparser as fp
url = "https://www.lemonde.fr/rss/une.xml" #"https://www.francetvinfo.fr/france.rss"
data = fp.parse(url)

# (c) visualiser les elements du fichier RSS
print(data.feed.title)
print(data.feed.published)

# (d) iterer sur les entrees du flux RSS
#for item in data.entries:
  #print(item.title, item.published, item.link, )

pprint(data.entries[0])

Le Monde.fr - Actualités et Infos en France et dans le monde
Thu, 02 Jan 2025 10:24:37 +0000
{'content': [{'base': 'https://www.lemonde.fr/rss/une.xml',
              'language': None,
              'type': 'text/plain',
              'value': 'Le président russe, Vladimir Poutine, le 26\xa0'
                       'décembre\xa02024\xa0dans la région de Leningrad.'}],
 'credit': 'Gavriil Grigorov / via REUTERS',
 'guidislink': True,
 'id': 'https://www.lemonde.fr/international/article/2025/01/02/en-russie-vladimir-poutine-affaibli-sur-le-front-economique-avant-le-retour-de-donald-trump_6477501_3210.html',
 'link': 'https://www.lemonde.fr/international/article/2025/01/02/en-russie-vladimir-poutine-affaibli-sur-le-front-economique-avant-le-retour-de-donald-trump_6477501_3210.html',
 'links': [{'href': 'https://www.lemonde.fr/international/article/2025/01/02/en-russie-vladimir-poutine-affaibli-sur-le-front-economique-avant-le-retour-de-donald-trump_6477501_3210.html',
            'rel': '

In [ ]:
import json
import ssl
import feedparser as fp
from pprint import pprint

# Itération sur les flux RSS
article_existants = list(france_tv_data.keys())
liste_flux = list(tv_info_source_data.values())
nouveaux_articles = []  # Pour stocker les nouveaux articles non déjà présents dans la base
for flux in liste_flux:
    data = fp.parse(flux)

    # Parcours des articles du flux
    for item in data.entries:
        # On suppose que l'identifiant unique est le lien (item.link)
        # Si l'article n'est pas déjà dans la base, on le stocke
        if item.link not in article_existants:
            # On peut récupérer diverses informations : titre, lien, date de publication, etc.
            # Selon la structure finale souhaitée, on ajoute l’article dans un format dict :
            nouvel_article = {
                "title": item.get("title", ""),
                "link": item.get("link", ""),
                "published": item.get("published", "")
            }
            nouveaux_articles.append(nouvel_article)


len(nouveaux_articles)
# Affichage des nouveaux articles trouvés
#print("Nouveaux articles trouvés :")
#for art in nouveaux_articles:
    #print(art["title"], art["published"], art["link"])




60

Il y a 60 nouveaux articles dans les flux RSS écoutés. Le dernier de la liste est titré **Rémunération, mobilité, flexibilité : qu’est-ce qui attend les '
          'cadres en 2025 ?** et publié le **Sat, 14 Dec 2024 10:26:51 +0100**.

In [ ]:
pprint(nouveaux_articles[59])

{'link': 'https://www.francetvinfo.fr/replay-radio/le-brief-eco/le-japonais-softbank-annonce-investir-cent-milliards-de-dollars-aux-etats-unis_6932525.html#xtor=RSS-3-[general]',
 'published': 'Tue, 17 Dec 2024 08:01:06 +0100',
 'title': 'Le japonais SoftBank annonce investir cent milliards de dollars aux '
          'États-Unis'}


# Question 2: Exploration du code HTML

- Le titre de l'article est contenu dans le \<head> entre les balises \<title>.

- Le texte de l'article se trouve dans le \<body>, réparti entre plusieurs balises \<p>. Ces balises contiennent les différents paragraphes du contenu principal. Toutefois, le texte d'un article peut aussi inclure des sous-titres ou d'autres éléments textuels importants qui ne se trouvent pas forcément dans des \<p>. Cela oblige à parcourir toute la structure HTML de la page pour extraire toutes les informations pertinentes sans inclure d'éléments indésirables comme des commentaires.

- Pour récupérer le texte, il faudrait analyser le HTML de la page,  en utilisant une bibliothèque comme BeautifulSoup. Cela permettrait d'identifier toutes les balises \<p> et d'extraire leur contenu. Chaque paragraphe extrait serait ensuite concaténé pour former le texte complet de l'article.

- La library newspaper agit comme un raccourci efficace pour obtenir les informations essentielles d'un article: titre, auteur, texte principal, etc.  Cette library simplifie et automatise les étapes que l'on ferait à la main avec BeautifulSoup. Avec newspaper, il n'est plus nécessaire de se soucier de la structure HTML spécifique au site web. Elle garantit une extractionstructurée et rapide.

On peut recupérer toutes les balises utilisées dans la page Html à l'aide de la fonction BeautifulSoup la méthode find_all. C'est ce qui est réalisé dans la cellule ci-dessous où l'on trouve entre autres les balises body, header, section etc.

In [ ]:
!pip install lxml[html_clean]
import json
import ssl
import feedparser as fp
import newspaper as np
from pprint import pprint
import pandas as pd
from bs4 import BeautifulSoup

url_article = "https://www.francetvinfo.fr/monde/europe/migrants/au-moins-cinq-morts-et-des-dizaines-de-disparus-apres-le-naufrage-d-un-bateau-de-migrants-en-grece_6955268.html#xtor=RSS-3-[general]"

article = np.Article(url_article)
article.download()
article.parse()

a = article.html


# Créer un objet BeautifulSoup pour analyser le HTML
soup = BeautifulSoup(a, "html.parser")

# Extraire tous les tags du document HTML
tags = {tag.name for tag in soup.find_all(True)}  # Utilisation d'un ensemble pour avoir des tags uniques

# Afficher les tags trouvés
print("Tags trouvés dans le document HTML :")
for tag in sorted(tags):
    print(tag)




Tags trouvés dans le document HTML :
a
article
aside
body
br
button
circle
defs
div
figcaption
figure
footer
form
h1
head
header
html
i
img
input
label
li
link
main
meta
nav
noscript
ol
p
path
picture
script
section
source
span
style
svg
symbol
template
time
title
ul
use


# Question 3: Mise à jour de la base de données d'articles
Ici nous mettons à jours la base de données avec le texte des articles du flux RSS. Nous avons fait le choix de conserver la même structure des article que celle de la base de données existante.

- 'author' : auteur de l'article,
- 'category': la catégorie de l'article (france, europe, économie, etc.)
- 'content' : Le contenue de l'article,
- 'date': date de publication de l'article,
- 'image_link': lien vers l'image d'illustration utilisé dans l'article,
- 'title': Titre de l'article

L'identification de ces nouveaux articles c'est faite par comparaison de leur lien url avec ceux des articles de la base de données.



In [ ]:
import json
import ssl
import feedparser as fp
import newspaper as np
from pprint import pprint
import pandas as pd

# (a) Prévenir les problèmes SSL
ssl._create_default_https_context = ssl._create_unverified_context

# Fonction pour mettre à jour la base de données avec les articles du flux RSS
def mettre_a_jour_base(flux_rss_dict, base_donnees):
    """
    Met à jour une base de données avec les articles récupérés depuis les flux RSS.

    :param flux_rss_dict: Dictionnaire avec les catégories comme clés et URLs RSS comme valeurs
    :param base_donnees: Dictionnaire contenant les articles existants (clé = lien)
    :return: (base mise à jour, articles ajoutés)
    """
    # Convertir les clés existantes en un ensemble pour éviter les doublons
    articles_existants = list(base_donnees.keys())

    # Nouveau dictionnaire pour les articles à ajouter
    nouveaux_articles = {}

    # Parcourir les flux RSS
    for category, url_flux in flux_rss_dict.items():
        print(f"Lecture du flux RSS '{category}' : {url_flux}")
        data = fp.parse(url_flux)

        # Parcourir chaque article du flux
        for item in data.entries:
            lien = item.get("link", "")
            if lien and lien not in articles_existants:
                try:
                    # Extraire les informations détaillées avec newspaper
                    article_obj = np.Article(lien)
                    article_obj.download()
                    article_obj.parse()

                    # Ajouter l'article au dictionnaire avec la clé 'category'
                    nouveaux_articles[lien] = {
                        "title": article_obj.title,
                        "author": article_obj.authors,
                        "date": article_obj.publish_date.isoformat(),
                        "content": article_obj.text,
                        "image_link": article_obj.top_image,
                        "category": category  # Ajouter la catégorie correspondant au flux
                    }
                    print(f"Article ajouté depuis '{category}': {lien}")
                except Exception as e:
                    print(f"Erreur pour l'article {lien} : {e}")

    # Mettre à jour la base de données existante
    base_donnees.update(nouveaux_articles)
    print(f"{len(nouveaux_articles)} articles ajoutés à la base.")
    return base_donnees, nouveaux_articles

a,b = mettre_a_jour_base(tv_info_source_data, france_tv_data)








Lecture du flux RSS 'france' : https://www.francetvinfo.fr/france.rss
Article ajouté depuis 'france': https://www.francetvinfo.fr/france/grand-est/bas-rhin/strasbourg/interpellation-d-un-homme-soupconne-d-avoir-renverse-mortellement-un-adolescent-de-15-ans-avec-sa-voiture-le-soir-du-nouvel-an-a-strasbourg_6990227.html#xtor=RSS-3-[general]
Article ajouté depuis 'france': https://www.francetvinfo.fr/faits-divers/accident/deux-randonneuses-meurent-en-montagne-apres-avoir-glisse-sur-plus-de-300-metres-en-haute-garonne_6989063.html#xtor=RSS-3-[general]
Article ajouté depuis 'france': https://www.francetvinfo.fr/france/auvergne-rhone-alpes/savoie/savoie-a-sainte-foy-tarentaise-ils-skient-en-slip-de-bain-pour-la-bonne-cause_6990200.html#xtor=RSS-3-[general]
Article ajouté depuis 'france': https://www.francetvinfo.fr/replay-radio/ca-dit-quoi/emeutes-en-nouvelle-caledonie-et-manifestations-contre-la-vie-chere-en-martinique-ca-dit-quoi_6986870.html#xtor=RSS-3-[general]
Article ajouté depuis 'fra

Au total il y a 60 nouveaux articles recupérés.

Nous en affichons les deux premiers de la liste.

In [ ]:
# Afficher les deux premiers articles ajoutés
for item in list(b.items())[:2]:
  pprint(item)

('https://www.francetvinfo.fr/france/grand-est/bas-rhin/strasbourg/interpellation-d-un-homme-soupconne-d-avoir-renverse-mortellement-un-adolescent-de-15-ans-avec-sa-voiture-le-soir-du-nouvel-an-a-strasbourg_6990227.html#xtor=RSS-3-[general]',
 {'author': ['Aurélien'],
  'category': 'france',
  'content': "Dans la nuit de mardi à mercredi, vers 1h du matin, l'accident "
             'mortel a eu lieu, rue de la Lauch.\n'
             '\n'
             "Un homme accusé d'avoir renversé mortellement un adolescent de "
             '15 ans avec sa voiture lors de la nuit du réveillon du Nouvel An '
             'a été interpellé, mercredi 1er janvier, au commissariat de '
             'Strasbourg où il était venu se rendre vers 22h30, a appris jeudi '
             'franceinfo de source policière. Cette source précise que cet '
             'homme est tout juste majeur.\n'
             '\n'
             'Dans la nuit de mardi à mercredi, vers 1h du matin à Strasbourg, '
             'un acc

In [ ]:
# Vérifier que les nouveaux articles ont tous été rajoutés
# La base initiale contient 2097 articles
len(france_tv_data)

2157

Avec 2156 articles, la base contient bien 60 articles de plus qu'initialement.

# Question 4: Sauvegarde locale des images

On pourra se servir de la library BeratifulSoup pour parcourir la structure HTML de la page de l'article et repérer la balise <img> contenant l'image. S'il y en a une, on recupère son lien dans le 'src' qui se trouve dans cette balise et on lance le téléchargement dans un répertoire locale créé à cet effet, tout en s'assurant d'avoir une nommenclature propre des fichiers qui permette d'éviter les conficts.

# Question 5: Sources d'informations complémentaires sur les articles

On pourrait extraire les commentaires, le temps de lecture estimé et les mots clés associés à l'article en utilisant BeautifulSoup pour le faire manuellement ou d'autres bibliothèques plus sophistiquées comme `Scrapy` pour des tâches de scraping plus complexes.

- **Commentaires** : On peut rechercher la balise `<section>` ayant l'id `comment` pour identifier les commentaires et les extraire sous forme d'une liste de dictionnaires comprenant les informations sur la date de publication du commentaire et les réponses faites aux commentaires. Par exemple :
  ```html
  <section id="comment">
    <div class="comment">
      <p class="comment-text">C'est triste cette tragedie</p>
      <span class="comment-date">2024-12-26</span>
    </div>
  </section>
  ```

- **Temps de lecture** : On peut identifier la balise `<span>` avec la classe `reading_time` pour extraire le temps de lecture estimé. Par exemple :
  ```html
  <span class="reading_time">Temps de lecture : 5 minutes</span>
  ```

- **Mots clés** : On peut localiser les mots clés dans les balises `<section>` avec la classe `related-tags`, les extraire avec les liens vers d'autres pages y étant relatifs. Par exemple :
  ```html
  <section class="related-tags">
    <a href="/tag/tech">Ouragan</a>
    <a href="/tag/science">Mayotte</a>
  </section>
  ```

# Question 7: Technique de détection des entités nommées

Les étiquettes IOB (Inside-Outside-Beginning) permettent de marquer chaque mot d’un texte en fonction de sa relation avec une entité nommée. Par exemple, dans la phrase "Emmanuel Macron est à Mayotte", "Emmanuel" et "Macron" pourraient être étiquetés respectivement comme B-PER (Beginning of a Person) et I-PER (Inside a Person), tandis que "Mayotte" serait étiqueté B-LOC (Beginning of a Location).

Les modèles de machine learning, comme les CRF (Conditional Random Fields) ou les réseaux LSTM , sont utilisés pour cette tâche en apprenant à prédire ces étiquettes à partir d’un ensemble de données étiquetées. Ces modèles utilisent des embeddings de mots pour mieux capturer le contexte et les relations entre les mots afin d'identifier avec précision les entités nommées dans un texte.

Une technique de détection des entités nommées consiste à attribuer à chaque mot de la phrase une étiquette de façon aléatoire et de mettre en place un modèle de RNN  qui va servir à prédire les différentes étiquettes pour les mots. Les modèles plus récents, comme les transformers (par exemple, BERT et GPT), ont une meilleure comprenhension du contexte global des phrases.

# Question 8: Détection des entités nommées

Ici nous commencer par charger le modèle fr_core_news_md pour le traitement du texte en français, et la pipeline senter est activé pour la segmentation des phrases.

La fonction extraire_entites_et_verbes définie prend en entrée la base de données des articles et retourne un dictionnaire d'articles enrichis avec les entités nommées et les verbes extraits du contenu des articles.

Pour chaque article dans base_articles :
Le contenu textuel de l'article est récupéré.
Si le texte est présent, il est traité avec le modèle nlp.


Les entités nommées sont extraites du texte et stockées dans une liste de dictionnaires, chaque dictionnaire contenant le texte de l'entité, son type, et ses positions de début et de fin dans le texte.

Les phrases sont extraites du contenu de l'article
et pour chaque phrase, les verbes sont identifiés et extraits sous forme de lemmes (formes de base des verbes).
Si une phrase contient des verbes ou des entités nommées, elle est ajoutée à la liste des phrases avec les informations sur les verbes et les entités.

- Construction du dictionnaire enrichi :

Pour chaque article, un dictionnaire est créé avec les informations suivantes : titre, auteur, date, contenu, lien de l'image, catégorie, entités nommées, et **phrases contenant des verbes et des entités**.

----
Nous avons constaté que la désactivation de certains traitements dans la chaîne de traitement de SpaCy, bien qu’ils ne soient pas directement utiles pour les tâches à accomplir, influence les sorties des entités au format IOB. Cela révèle que les différents traitements de la méthode nlp ne sont pas totalement indépendants, ce qui a motivé notre choix de conserver l’ensemble des composants activés.

Par exemple, en désactivant le module d’embeddings, l’entité "Emmanuel Macron" n’est plus reconnue comme une entité unique. À la place, deux entités distinctes, "Emmanuel" et "Macron", sont restituées. Ce résultat met en évidence une perte d’information, car le contexte global du texte n’est plus suffisamment pris en compte dans les sorties. Cette observation souligne l’importance d’une approche contextuelle pour préserver la cohérence et la précision des entités identifiées.

In [ ]:
import json
import spacy


# Charger le modèle sans rien désactiver
nlp = spacy.load("fr_core_news_md")

# Activer le senter pour la segmentation
nlp.enable_pipe("senter")

def extraire_entites_et_verbes(base_articles):

  """
  Fonction pour extraire les entités nommées et les verbes d'un ensemble d'articles
  Entrée : base_articles - dictionnaire contenant les articles avec leur contenu textuel
  Sortie : dictionnaire enrichi avec les entités nommées et les verbes extraits

  """
  articles_enrichis = {}
  for lien, article in base_articles.items():
      texte = article.get("content", "")
      if texte:
          doc = nlp(texte)

          # Extraire les entités nommées
          entites = [
                {"entite": ent.text, "type": ent.label_, "start": ent.start_char, "end": ent.end_char}
                for ent in doc.ents
            ]

          # Extraire les phrases
          phrases = []
          for sent in doc.sents:
              # Extraire les verbes
              verbes = [
                    {"verbe": token.lemma_, "start": token.idx, "end": token.idx + len(token)}
                    for token in sent if token.pos_ == "VERB"
                ]
              # Ajout de la phrase si verbes ou entités
              if verbes or sent.ents:
                  phrases.append({
                        "texte": sent.text.strip(),
                        "start": sent.start_char,
                        "end": sent.end_char,
                        "verbes": verbes,
                        "entites": [
                            {
                                "entite": ent.text,
                                "type": ent.label_,
                                "start": ent.start_char,
                                "end": ent.end_char
                            }
                            for ent in sent.ents
                        ]
                    })

          articles_enrichis[lien] = {
                "title": article.get("title", ""),
                "author": article.get("author", ""),
                "date": article.get("date", ""),
                "content": texte,
                "image_link": article.get("image_link", ""),
                "category": article.get("category", ""),
                "entities": entites,
                "phrases": phrases
            }
  return articles_enrichis





#test = first_three_elements = dict(list(france_tv_data.items())[:3])
#article_enrichis_test = extraire_entites_et_verbes(test)
articles_enrichis = extraire_entites_et_verbes(france_tv_data)


visualiser les informations extraites des articles. Nous avons utilisé une boucle afin d'itéré sur plus d'articles au besoin.

Cas du premier article:

In [ ]:
from pprint import pprint

# Limiter l'affichage à 5 articles
for lien, item in list(articles_enrichis.items())[:1]:
    print(f"Lien : {lien}")
    print("Phrases et leurs éléments :")

    # Parcourir et afficher chaque phrase avec ses éléments
    for phrase in item.get("phrases", []):
        print(f"Phrase : {phrase['texte']}")
        print(f" - Indices : start={phrase['start']}, end={phrase['end']}")

        # Afficher les verbes
        if phrase.get("verbes"):
            print(" - Verbes :")
            for verbe in phrase["verbes"]:
                print(f"   * {verbe['verbe']} (start={verbe['start']}, end={verbe['end']})")
        else:
            print(" - Verbes : Aucun")

        # Afficher les entités
        if phrase.get("entites"):
            print(" - Entités :")
            for entite in phrase["entites"]:
                print(f"   * {entite['entite']} ({entite['type']}) - start={entite['start']}, end={entite['end']}")
        else:
            print(" - Entités : Aucune")

        print("-" * 50)  # Séparateur entre phrases

    print("=" * 50)  # Séparateur entre articles


Lien : https://www.francetvinfo.fr/sante/prevention/eure-quand-les-elus-apprennent-a-se-defendre_3791635.html#xtor=RSS-3-[france]
Phrases et leurs éléments :
Phrase : Beaucoup de maires et de secrétaires de mairie ont rencontré au cours de leurs mandats des problèmes de violence, essentiellement verbale.
 - Indices : start=0, end=138
 - Verbes :
   * rencontrer (start=51, end=60)
 - Entités : Aucune
--------------------------------------------------
Phrase : C'est le cas de Claire.
 - Indices : start=139, end=162
 - Verbes : Aucun
 - Entités :
   * Claire (PER) - start=155, end=161
--------------------------------------------------
Phrase : Un jour, un homme est entré dans son bureau et a frappé très fort sur son bureau pour l'intimider.
 - Indices : start=163, end=261
 - Verbes :
   * entrer (start=185, end=190)
   * frapper (start=212, end=218)
 - Entités : Aucune
--------------------------------------------------
Phrase : "J'ai réagi à l'instinct, j'ai essayé de discuter, mais ça s'

In [ ]:
for lien, item in list(articles_enrichis.items())[:1]:
    print(f"Lien : {lien}")
    pprint(item)
    print("=" * 50)  # Séparateur pour plus de lisibilité

Lien : https://www.francetvinfo.fr/sante/prevention/eure-quand-les-elus-apprennent-a-se-defendre_3791635.html#xtor=RSS-3-[france]
{'author': [],
 'category': 'france',
 'content': 'Beaucoup de maires et de secrétaires de mairie ont rencontré au '
            'cours de leurs mandats des problèmes de violence, essentiellement '
            "verbale. C'est le cas de Claire. Un jour, un homme est entré dans "
            "son bureau et a frappé très fort sur son bureau pour l'intimider. "
            '"J\'ai réagi à l\'instinct, j\'ai essayé de discuter, mais ça '
            "s'est envenimé. Heureusement qu'une dame est arrivée à la "
            'permanence et a menacé d\'appeler la gendarmerie", raconte Claire '
            'Carrère-Godebout, maire (SE) de Graveron-Sémerville (Eure).\n'
            '\n'
            'Des professeurs de karaté et un psychologue\n'
            '\n'
            'La plupart de ces élus ne savent pas comment se comporter face à '
            "une situation co

# Question 9: les 20 entités les plus fréquentes dans la collection de documents (par types d'entités)

Les fonctions ci-dessous ont pour objectif d'analyser les entités nommées dans un ensemble d'articles et de déterminer les 20 entités les plus fréquentes pour chaque type d'entité (personnes, lieux, organisations).

In [ ]:
import json
from collections import Counter

# Fonction pour déterminer les 20 entités les plus fréquentes par type
def top_entites_par_type(fichier):
    """
    Analyse les entités nommées et détermine les 20 plus fréquentes par type.
    :param fichier_json: Chemin vers le fichier JSON contenant les entités enrichies.
    :return: Trois listes avec les entités les plus fréquentes pour PER, LOC, et ORG.
    """
    # Compteurs pour chaque type d'entité
    compteur_per = Counter()
    compteur_loc = Counter()
    compteur_org = Counter()

    # Charger les données enrichies

    base_enrichie = fichier

    # Parcourir chaque article pour collecter les entités
    for article in base_enrichie.values():
        for entite in article.get("entities", []):
            if entite["type"] == "PER":
                compteur_per[entite["entite"]] += 1
            elif entite["type"] == "LOC":
                compteur_loc[entite["entite"]] += 1
            elif entite["type"] == "ORG":
                compteur_org[entite["entite"]] += 1

    # Obtenir les 20 entités les plus fréquentes pour chaque type
    top_per = compteur_per.most_common(20)
    top_loc = compteur_loc.most_common(20)
    top_org = compteur_org.most_common(20)

    return top_per, top_loc, top_org


def afficher_top_entites(top_per, top_loc, top_org):
    """
    Affiche proprement les entités les plus fréquentes pour chaque type (PER, LOC, ORG).
    :param top_per: Liste des entités 'PER' (Personnes) les plus fréquentes.
    :param top_loc: Liste des entités 'LOC' (Lieux) les plus fréquentes.
    :param top_org: Liste des entités 'ORG' (Organisations) les plus fréquentes.
    """
    print("\n" + "=" * 50)
    print(f"{'Top 20 des entités de type PER (Personnes)':^50}")
    print("=" * 50)
    for i, (entite, freq) in enumerate(top_per, start=1):
        print(f"{i:>2}. {entite:<30} - {freq} occurrence(s)")

    print("\n" + "=" * 50)
    print(f"{'Top 20 des entités de type LOC (Lieux)':^50}")
    print("=" * 50)
    for i, (entite, freq) in enumerate(top_loc, start=1):
        print(f"{i:>2}. {entite:<30} - {freq} occurrence(s)")

    print("\n" + "=" * 50)
    print(f"{'Top 20 des entités de type ORG (Organisations)':^50}")
    print("=" * 50)
    for i, (entite, freq) in enumerate(top_org, start=1):
        print(f"{i:>2}. {entite:<30} - {freq} occurrence(s)")








In [ ]:
a,b,c = top_entites_par_type(articles_enrichis)
afficher_top_entites(a,b,c)


    Top 20 des entités de type PER (Personnes)    
 1. Emmanuel Macron                - 258 occurrence(s)
 2. Covid-19                       - 182 occurrence(s)
 3. Angela Merkel                  - 118 occurrence(s)
 4. Bruno Le Maire                 - 115 occurrence(s)
 5. Boris Johnson                  - 68 occurrence(s)
 6. Muriel Pénicaud                - 59 occurrence(s)
 7. Donald Trump                   - 52 occurrence(s)
 8. Laurent Desbonnets             - 50 occurrence(s)
 9. président de la République     - 48 occurrence(s)
10. Thierry Breton                 - 44 occurrence(s)
11. Ursula von der Leyen           - 41 occurrence(s)
12. Edouard Philippe               - 40 occurrence(s)
13. Birkenau                       - 37 occurrence(s)
14. Vladimir Poutine               - 33 occurrence(s)
15. Olivier Véran                  - 33 occurrence(s)
16. Recep Tayyip Erdogan           - 32 occurrence(s)
17. Covid                          - 29 occurrence(s)
18. Kyriakos Mitsotakis   

**Entités de type PER (Personnes)**
- Emmanuel Macron (256 occurrences) : En tant que président de la République française, il est souvent mentionné dans les articles.
- Covid-19 (183 occurrences) : Bien que ce soit une maladie, elle est identifié comme une entité nommée dans les textes.
- Angela Merkel (118 occurrences) :  Ancienne chancelière allemande, elle reste une figure politique importante.
- Bruno Le Maire (115 occurrences) : Ministre français de l'Économie, des Finances et de la Relance.
- Boris Johnson (68 occurrences) : Ancien Premier ministre du Royaume-Uni, souvent mentionné dans le contexte du Brexit et de la politique britannique.

Les autres noms incluent des figures politiques et publiques importantes comme Muriel Pénicaud, Thierry Breton, Ursula von der Leyen, Donald Trump, et Vladimir Poutine.
La présence de George Floyd (27 occurrences) reflète l'impact mondial de son décès et des mouvements sociaux qui ont suivi.

**Entités de type LOC (Lieux)**
- France (825 occurrences) : Le pays est naturellement le plus mentionné, étant le contexte principal des articles.
- Allemagne (571 occurrences) : En tant que principal partenaire européen de la France, elle est fréquemment mentionnée.
- Paris (396 occurrences) : La capitale française est un centre politique, économique et culturel majeur.
- D'autres lieux importants incluent Europe, Royaume-Uni, Italie, Chine, Grèce, et Turquie.
- La mention de Franceinfo (114 occurrences) comme lieu peut être due à une erreur de classification ou à un contexte spécifique dans les articles.

**Entités de type ORG (Organisations)**
- JT (2082 occurrences) : Les journaux télévisés (JT) sont souvent cités comme sources d'information.
- franceinfo (509 occurrences) : La chaîne d'information continue est une source majeure de nouvelles.
- Covid-19 (454 occurrences) : Encore une fois, bien que ce soit une maladie, elle est traitée comme une entité organisationnelle dans certains contextes.
- D'autres organisations importantes incluent Union européenne, AFP, France Télévisions, Commission européenne, Renault, et Amazon.
- La mention de Brexit (134 occurrences) comme organisation peut être due à une classification contextuelle spécifique.

De façon générale, Covid-19 apparaît dans plusieurs catégories, ce qui montre son omniprésence dans les discussions et les articles.
Les entités politiques et économiques dominent les mentions, reflétant l'importance de ces sujets dans les médias.
Certaines entités peuvent être mal classifiées (comme Franceinfo en tant que lieu), ce qui pourrait nécessiter une revue manuelle ou des ajustements dans le modèle de classification.
Ces résultats fournissent une vue d'ensemble des sujets et des entités les plus discutés dans les articles analysés, offrant des insights précieux pour des analyses plus approfondies. Si vous avez des questions ou besoin de précisions supplémentaires, n'hésitez pas à demander !

# Question 10: Co-occurence d'entités au sein du même document

Cette fonction permet d'analyser les relations entre différentes entités dans un corpus de textes. Par exemple, elle peut aider à comprendre combien de fois deux personnalités politiques sont mentionnées ensemble dans les articles, ou combien de fois une entreprise et un lieu sont cités conjointement

In [ ]:
import json

# Fonction pour compter les co-occurrences de deux entités
def compter_cooccurrences(base_donnees, entite1, type1, entite2, type2):
    """
    Compte le nombre de co-occurrences des deux entités dans la même phrase ou document.

    :param base_donnees: Dictionnaire contenant les articles enrichis avec leurs entités.
    :param entite1: Texte de la première entité.
    :param type1: Type de la première entité (ex : 'PER', 'LOC').
    :param entite2: Texte de la deuxième entité.
    :param type2: Type de la deuxième entité (ex : 'LOC', 'ORG').
    :return: Nombre de documents où les deux entités co-occurrent.
    """
    cooccurrence_count = 0

    # Parcourir chaque article dans la base
    for lien, article in base_donnees.items():
        entites = article.get("entities", [])

        # Filtrer les entités correspondantes à entite1 et entite2
        presence_entite1 = any(ent["entite"] == entite1 and ent["type"] == type1 for ent in entites)
        presence_entite2 = any(ent["entite"] == entite2 and ent["type"] == type2 for ent in entites)

        # Si les deux entités sont présentes, on incrémente le compteur
        if presence_entite1 and presence_entite2:
            cooccurrence_count += 1

    return cooccurrence_count



In [ ]:
entite1, type1 = "Emmanuel Macron", "PER"
entite2, type2 = "France", "LOC"

# Appliquer la fonction pour compter les co-occurrences
resultat = compter_cooccurrences(articles_enrichis, entite1, type1, entite2, type2)

# Afficher le résultat
print(f"Nombre de documents où '{entite1}' ({type1}) et '{entite2}' ({type2}) co-occurrent : {resultat}")

Nombre de documents où 'Emmanuel Macron' (PER) et 'France' (LOC) co-occurrent : 47


# Question 11: Co-occurences d'entités au sein du même document (les plus fréquentes)

Cette fonction peut être utilisée pour analyser les relations entre différentes entités dans un corpus de textes. Par exemple, elle peut aider à comprendre quelles personnalités politiques et lieux sont souvent mentionnés ensemble, ou quelles organisations et lieux sont fréquemment cités conjointement.

In [ ]:

from itertools import combinations
from itertools import product
# Fonction pour trouver les paires d'entités de types différents les plus fréquentes
def paires_top_entites(base_donnees, top_per, top_loc, top_org):
    """
    Trouve les paires d'entités (types différents) parmi les top 20 entités par type.
    """
    compteur_paires = Counter()

    # Extraire les noms des entités à partir des top listes
    top_personnes = [entite for entite, _ in top_per]
    top_lieux = [entite for entite, _ in top_loc]
    top_organisations = [entite for entite, _ in top_org]

    # Créer des combinaisons possibles (PER-LOC, PER-ORG, LOC-ORG)
    paires_a_tester = list(product(top_personnes, ["PER"], top_lieux, ["LOC"])) + \
                      list(product(top_personnes, ["PER"], top_organisations, ["ORG"])) + \
                      list(product(top_lieux, ["LOC"], top_organisations, ["ORG"]))

    # Compter les co-occurrences pour chaque paire
    for entite1, type1, entite2, type2 in paires_a_tester:
        cooccurrences = compter_cooccurrences(base_donnees, entite1, type1, entite2, type2)
        if cooccurrences > 0:
            paire = tuple(sorted([(entite1, type1), (entite2, type2)]))
            compteur_paires[paire] = cooccurrences

    return compteur_paires.most_common(10)



Cette fonction est utilisée pour afficher les résultats des paires d'entités les plus fréquentes trouvées par la fonction précédente (paires_top_entites). Elle permet de visualiser facilement quelles paires d'entités de types différents apparaissent le plus souvent ensemble dans les articles analysés.

In [ ]:
def afficher_top_paires(top_paires):
    """
    Affiche proprement les paires d'entités avec leurs types et fréquences.
    :param top_paires: Liste des paires d'entités et leurs fréquences.
    """
    print("\n" + "=" * 60)
    print(f"{'Top 10 des paires d entités de types différents':^60}")
    print("=" * 60)

    print(f"{'N°':<4} {'Entité 1':<25} {'Type 1':<8} | {'Entité 2':<25} {'Type 2':<8} | {'Occurrences':<10}")
    print("-" * 60)

    # Parcourir et afficher chaque paire avec son numéro
    for idx, (paire, freq) in enumerate(top_paires, start=1):
        entite1, type1 = paire[0]
        entite2, type2 = paire[1]
        print(f"{idx:<4} {entite1:<25} {type1:<8} | {entite2:<25} {type2:<8} | {freq:<10}")

    print("=" * 60)


In [ ]:
top_paires = paires_top_entites(articles_enrichis, a, b, c)
afficher_top_paires(top_paires)


      Top 10 des paires d entités de types différents       
N°   Entité 1                  Type 1   | Entité 2                  Type 2   | Occurrences
------------------------------------------------------------
1    France                    LOC      | JT                        ORG      | 204       
2    Allemagne                 LOC      | JT                        ORG      | 160       
3    JT                        ORG      | Paris                     LOC      | 110       
4    Covid-19                  ORG      | France                    LOC      | 107       
5    JT                        ORG      | la France                 LOC      | 107       
6    France                    LOC      | franceinfo                ORG      | 100       
7    Français                  LOC      | JT                        ORG      | 97        
8    JT                        ORG      | État                      LOC      | 80        
9    Italie                    LOC      | JT                      

**Dominance des médias**

- JT (Journal Télévisé) apparaît dans 7 des 10 paires les plus fréquentes, souvent associé à des lieux comme France, Allemagne, Paris, et Italie. Cela montre l'importance des journaux télévisés dans la couverture des événements nationaux et internationaux.
Franceinfo et France Télévisions sont également présents, soulignant le rôle central des médias français dans la diffusion de l'information.

**Importance des lieux**

- France est l'entité la plus fréquemment associée, apparaissant dans plusieurs paires avec des organisations comme JT, franceinfo, et Covid-19. Cela reflète l'importance de la France dans les articles analysés.
- Allemagne, Paris, et Italie sont également des lieux fréquemment mentionnés, souvent en relation avec des médias.

**Impact de la pandémie**

Covid-19 est associé à France dans une paire fréquente, ce qui montre l'importance continue de la pandémie dans les discussions médiatiques.


# Questions 12: Co-occurence d'entités au sein de la même phrase (les plus fréquents)

In [ ]:
from collections import Counter
from itertools import product

def compter_cooccurrences_phrase(base_donnees, entite1, type1, entite2, type2):
    """
    Compte le nombre de phrases contenant les deux entités dans un article.
    :param base_donnees: Dictionnaire contenant les articles enrichis avec phrases et entités.
    :param entite1: Texte de la première entité.
    :param type1: Type de la première entité (ex : 'PER', 'LOC').
    :param entite2: Texte de la deuxième entité.
    :param type2: Type de la deuxième entité (ex : 'LOC', 'ORG').
    :return: Nombre de phrases où les deux entités co-occurrent.
    """
    cooccurrence_count = 0

    # Parcourir chaque article
    for article in base_donnees.values():
        phrases = article.get("phrases", [])

        # Parcourir chaque phrase et vérifier la présence des entités
        for phrase in phrases:
            ents_in_phrase = phrase.get("entites", [])
            presence_entite1 = any(ent["entite"] == entite1 and ent["type"] == type1 for ent in ents_in_phrase)
            presence_entite2 = any(ent["entite"] == entite2 and ent["type"] == type2 for ent in ents_in_phrase)

            # Compter si les deux entités sont dans la même phrase
            if presence_entite1 and presence_entite2:
                cooccurrence_count += 1

    return cooccurrence_count

def paires_top_entites_par_phrase(base_donnees, top_per, top_loc, top_org):
    """
    Trouve les paires d'entités co-occurrentes dans les mêmes phrases.
    """
    compteur_paires = Counter()

    # Extraire les noms des entités des top listes
    top_personnes = [entite for entite, _ in top_per]
    top_lieux = [entite for entite, _ in top_loc]
    top_organisations = [entite for entite, _ in top_org]

    # Créer des combinaisons possibles (PER-LOC, PER-ORG, LOC-ORG)
    paires_a_tester = list(product(top_personnes, ["PER"], top_lieux, ["LOC"])) + \
                      list(product(top_personnes, ["PER"], top_organisations, ["ORG"])) + \
                      list(product(top_lieux, ["LOC"], top_organisations, ["ORG"]))

    # Compter les co-occurrences pour chaque paire dans les phrases
    for entite1, type1, entite2, type2 in paires_a_tester:
        cooccurrences = compter_cooccurrences_phrase(base_donnees, entite1, type1, entite2, type2)
        if cooccurrences > 0:
            paire = tuple(sorted([(entite1, type1), (entite2, type2)]))
            compteur_paires[paire] = cooccurrences

    return compteur_paires.most_common(10)

def afficher_top_paires_par_phrase(top_paires):
    """
    Affiche proprement les paires d'entités co-occurrentes dans les mêmes phrases.
    """
    print("\n" + "=" * 60)
    print(f"{'Top 10 des paires d entités dans une même phrase':^60}")
    print("=" * 60)

    print(f"{'N°':<4} {'Entité 1':<25} {'Type 1':<8} | {'Entité 2':<25} {'Type 2':<8} | {'Occurrences':<10}")
    print("-" * 60)

    # Parcourir et afficher chaque paire avec son numéro
    for idx, (paire, freq) in enumerate(top_paires, start=1):
        entite1, type1 = paire[0]
        entite2, type2 = paire[1]
        print(f"{idx:<4} {entite1:<25} {type1:<8} | {entite2:<25} {type2:<8} | {freq:<10}")

    print("=" * 60)


In [ ]:
# Exemple d'utilisation avec les top listes (top_per, top_loc, top_org) déjà calculées
top_paires_phrase = paires_top_entites_par_phrase(articles_enrichis, a, b, c)
afficher_top_paires_par_phrase(top_paires_phrase)


      Top 10 des paires d entités dans une même phrase      
N°   Entité 1                  Type 1   | Entité 2                  Type 2   | Occurrences
------------------------------------------------------------
1    France Télévisions        ORG      | Franceinfo                LOC      | 67        
2    Franceinfo                LOC      | Magazines                 ORG      | 66        
3    Royaume-Uni               LOC      | Union européenne          ORG      | 54        
4    Covid-19                  ORG      | France                    LOC      | 39        
5    Royaume-Uni               LOC      | UE                        ORG      | 39        
6    Berlin                    LOC      | Laurent Desbonnets        PER      | 37        
7    Muriel Pénicaud           PER      | Travail                   ORG      | 29        
8    Commission européenne     ORG      | Ursula von der Leyen      PER      | 24        
9    Brexit                    ORG      | Royaume-Uni             

- France Télévisions (ORG) - Franceinfo (LOC) (67 occurrences) et Franceinfo (LOC) - Magazines (ORG) (66 occurrences) :
indiquent une forte relation entre ces deux entités médiatiques, probablement en raison de citations ou de collaborations fréquentes.


- Royaume-Uni (LOC) - Union européenne (ORG) (54 occurrences), Royaume-Uni (LOC) - UE (ORG) (39 occurrences) et Brexit (ORG) - Royaume-Uni (LOC) (23 occurrences): reflètent les discussions fréquentes sur les relations post-Brexit entre le Royaume-Uni et l'UE.

- Covid-19 (ORG) - France (LOC) (39 occurrences) :
La pandémie de Covid-19 et son impact en France restent des sujets majeurs dans les discussions.

- Berlin (LOC) - Laurent Desbonnets (PER) (37 occurrences) :
 pourrait indiquer des articles spécifiques impliquant Laurent Desbonnets et la ville de Berlin.

- Muriel Pénicaud (PER) - Travail (ORG) (29 occurrences) :
En tant qu'ancienne ministre du Travail, Muriel Pénicaud est souvent mentionnée en relation avec des sujets de travail et d'emploi.

- Commission européenne (ORG) - Ursula von der Leyen (PER) (24 occurrences) :
Ursula von der Leyen, en tant que présidente de la Commission européenne, est fréquemment associée à cette organisation.

- Olivier Véran (PER) - ministre de la Santé (ORG) (19 occurrences) :
Olivier Véran, en tant que ministre de la Santé, est naturellement associé à ce titre dans les discussions.

**Conclusion**

Ces résultats montrent l'importance des médias dans la couverture des événements, des relations internationales, et des personnalités politiques. Les discussions sur le Brexit, la pandémie de Covid-19, et les relations entre le Royaume-Uni et l'UE sont particulièrement fréquentes. Les personnalités politiques et les organisations jouent également un rôle central dans les discussions.

# Question 13: Caractéristion d'une paire d'entités au sein d'une même phrase par des verbes

Nous utiliserons la base d'articles enrichie à la question 8. Cette base contient, pour chaque article, les entités nommées des phrases comportant au moins un verbe, ainsi que la liste des verbes associés. Cela permet d'éviter les extractions répétées d'entités et de verbes, rendant l'analyse plus rapide et intuitive.

Cependant, plusieurs verbes peuvent apparaître entre deux entités dans une même phrase, ce qui nécessite une analyse plus poussée pour identifier les verbes qui traduisent réellement la relation entre elles. Cette analyse peut être réalisée en ajustant partiellement notre code pour recupérer l'arbre de dépendance syntaxique.

In [ ]:
def extraire_verbes_entre_entites(base_articles, entite1, type1, entite2, type2):
    """
    Trouve les verbes situés entre deux entités spécifiées dans chaque phrase.
    :param base_articles: Dictionnaire contenant les articles enrichis.
    :param entite1: Texte de la première entité.
    :param type1: Type de la première entité (PER, LOC, ORG, etc.).
    :param entite2: Texte de la deuxième entité.
    :param type2: Type de la deuxième entité.
    :return: Liste des verbes lemmatisés entre les deux entités.
    """
    relations = []

    # Parcourir chaque article
    for lien, article in base_articles.items():
        phrases = article.get("phrases", [])

        # Parcourir chaque phrase et vérifier la présence des entités
        for phrase in phrases:
          ents_in_phrase = phrase.get("entites", [])
          verbs_in_phrase = phrase.get("verbes",[])

          # Identifier les deux entités spécifiées
          e1 = next((ent for ent in ents_in_phrase if ent["entite"] == entite1 and ent["type"] == type1),None)
          e2 = next((ent for ent in ents_in_phrase if ent["entite"] == entite2 and ent["type"] == type2),None)

          if e1 and e2:
            # Définir les positions des deux entités
            if e1["start"] < e2["start"]:
              start, end = e1["end"], e2["start"]
            else:
              start, end = e2["end"], e1["start"]
            #start, end = sorted([e1["start"], e2["start"]])

            # Extraire les verbes entre les deux entités
            verbes = [verbe["verbe"] for verbe in verbs_in_phrase if verbe["start"] > start and verbe["end"] < end]
            if verbes:
              relations.append({
                  "phrase": phrase.get("texte"),
                  "verbes": verbes,
                  "article": lien
                        })
    return relations


# Question 14: Application de la caractéristion par des verbes pour des paires d'entités fréquentes

Dans cette partie nous appliquons la fonction de la question 13 à deux paires d'entités fréquents obtenus à la question 12: Royaume-Uni (LOC) / Union européenne (ORG), Berlin (LOC) / Laurent Desbonnets (PER).

Pour s'amuser nous avons appliquer cela à la paire Mayotte (LOC) / Emmanuel Macron (PER).

## Application à la paire Royaume-Uni (LOC) / Union européenne (ORG)

In [ ]:
# Exemple d'utilisation

entite1, type1 = "Royaume-Uni", "LOC"
entite2, type2 = "Union européenne", "ORG"
relations = extraire_verbes_entre_entites(articles_enrichis, entite1, type1, entite2, type2)

for relation in relations:
  print(f"Article : {relation['article']}")
  print(f"Phrase  : {relation['phrase']}")
  print(f"Verbes  : {', '.join(relation['verbes'])}")
  print("-" * 80)


Article : https://www.francetvinfo.fr/monde/europe/la-grande-bretagne-et-l-ue/brexit-boris-johnson-a-signe-l-accord-de-retrait-de-l-ue_3798791.html#xtor=RSS-3-[monde/europe]
Phrase  : Le Royaume-Uni pourra alors sortir de l'Union européenne le 31 janvier, après 47 ans d'un mariage qui se sera achevé par plus de trois ans de négociations difficiles et de débats acrimonieux côté britannique sur les modalités du divorce.
Verbes  : pouvoir, sortir
--------------------------------------------------------------------------------
Article : https://www.francetvinfo.fr/replay-radio/on-ne-pouvait-pas-le-rater/on-ne-pouvait-pas-le-rater-derniere-bataille-en-musique-pour-les-pro-et-anti-brexit_3785317.html#xtor=RSS-3-[monde/europe]
Phrase  : Le Premier ministre Boris Johnson dans son discours devant le Parti conservateur àManchester (Grande-Bretagne), le 2 octobre 2019. (BEN STANSALL / AFP)

Vendredi 31 janvier, le Royaume-Uni sortira officiellement de l'Union européenne.
Verbes  : sortir
--------

On observe que le verbe "quitter" revient fréquemment dans les relations entre le Royaume-Uni et l'Union européenne. Cette récurrence reflète naturellement le contexte du Brexit, marqué par la sortie officielle du Royaume-Uni de l'Union européenne.

En outre, d'autres verbes, tels que "suspendre" et "exporter", témoignent des dossiers de négociation les plus houleux ayant marqué cette période. Par exemple, "suspendre" est lié aux tensions autour des droits de pêche, tandis que "exporter" fait référence aux débats sur les relations commerciales post-Brexit.

-----
Cependant, il est également apparu que l'extraction des phrases présente certaines limitations. Notamment, des fragments issus de sous-titres ou d'autres éléments non pertinents des articles ont été inclus par erreur. Cela est particulièrement visible dans l'exemple suivant :

Article :
[Dordogne : le Brexit vu du village le plus anglais de France](https://www.francetvinfo.fr/monde/europe/la-grande-bretagne-et-l-ue/dordogne-le-brexit-vu-du-village-le-plus-anglais-de-france_3807787.html#xtor=RSS-3-[france])

Phrase extraite :
"De nombreuses peurs. Ils ne peuvent pas cacher leur peine le jour où le Royaume-Uni quitte l'Union européenne."
Problème :
"De nombreuses peurs" est un sous-titre et ne devrait pas apparaître dans l'extraction.
Malgré ces limitations, l'extraction des verbes entre les entités identifiées fonctionne globalement bien et permet de dégager des relations cohérentes et pertinentes.

## Application à la paire Berlin (LOC) / Laurent Desbonnets (PER)

In [ ]:
# Exemple d'utilisation

entite1, type1 = "Laurent Desbonnets", "PER"
entite2, type2 = "Berlin", "LOC"
relations = extraire_verbes_entre_entites(articles_enrichis, entite1, type1, entite2, type2)

for relation in relations:
  print(f"Article : {relation['article']}")
  print(f"Phrase  : {relation['phrase']}")
  print(f"Verbes  : {', '.join(relation['verbes'])}")
  print("-" * 80)

Article : https://www.francetvinfo.fr/monde/europe/allemagne/charbon-les-allemands-dependent-encore-des-centrales-a-charbon_3804519.html#xtor=RSS-3-[monde/europe]
Phrase  : "Même sans se chauffer au charbon, presque tous les Allemands en consomment tous les jours, par exemple en allumant la lumière ou en branchant un appareil, car en Allemagne 40% de l'électricité provient encore de centrales à charbon", conclut Laurent Desbonnets, correspondant à Berlin.
Verbes  : correspondre
--------------------------------------------------------------------------------
Article : https://www.francetvinfo.fr/sante/maladie/coronavirus/face-a-l-epidemie-de-coronavirus-la-frontiere-franco-allemande-partiellement-fermee_3869723.html#xtor=RSS-3-[monde/europe]
Phrase  : C'est exceptionnel", commente lundi matin Laurent Desbonnets, correspondant de France Télévisions à Berlin.
Verbes  : correspondre
--------------------------------------------------------------------------------
Article : https://www.franc

Ce cas met clairement en évidence que Laurent Desbonnets joue le rôle d'un acteur de liaison avec Berlin, étant donné qu'il est régulièrement présenté comme un journaliste correspondant basé dans cette ville. En examinant les phrases où ces deux entités apparaissent conjointement, il devient évident que son rôle consiste à rapporter et commenter les événements locaux ou internationaux depuis Berlin.

-----
Cependant, une anomalie dans l'analyse grammaticale apparaît concernant le mot "correspondant". Dans les phrases étudiées, ce terme est utilisé comme un nom commun, désignant la fonction journalistique de Laurent Desbonnets. Malgré cela, il a été identifié comme un verbe par l'outil d'analyse linguistique (part-of-speech tagging). Cette confusion entraîne une classification incorrecte et pourrait fausser l'interprétation des relations entre les entités dans le contexte.

## Application à la paire Mayotte (LOC) / Emmanuel Macron (PER)

In [ ]:
# Exemple d'utilisation

entite1, type1 = "Mayotte", "LOC"
entite2, type2 = "Emmanuel Macron", "PER"
relations = extraire_verbes_entre_entites(articles_enrichis, entite1, type1, entite2, type2)

for relation in relations:
  print(f"Article : {relation['article']}")
  print(f"Phrase  : {relation['phrase']}")
  print(f"Verbes  : {', '.join(relation['verbes'])}")
  print("-" * 80)


Article : https://www.francetvinfo.fr/replay-radio/les-informes-du-matin/la-relation-macron-bayrou-l-appel-du-patronat-et-des-syndicats-les-informes-du-matin-du-mercredi-18-decembre_6934433.html#xtor=RSS-3-[general]
Phrase  : Emmanuel Macron est en effet attendu mercredi à Bruxelles avant de s'envoler pour Mayotte.
Verbes  : attendre, envoler
--------------------------------------------------------------------------------


La relation entre la visite d'Emmanuel Macron à Mayotte et le passage du cyclone Chido est clairement exprimée par le verbe "envoler", qui marque un lien direct.

----
En revanche, "attendre" traduit seulement une relation géographique entre Macron et Bruxelles, sans interaction directe entre entités. Cela suggère qu'une analyse approfondie des verbes, notamment via un arbre de dépendance syntaxique, permettrait de mieux identifier ceux qui reflètent réellement des relations entre les entités impliquées.

# Question 15: Utilité  des couples et verbes caractéristiques

- Les couples entités-verbes Royaume-Uni - quitter - Union européenne ou Royaume-Uni - négocier - accords commerciaux, permettent de mieux comprendre les relations politiques, économiques et historiques du Brexit. Ils peuvent révéler de nouvelles connexions potentielles et de regrouper les relations similaires pour une analyse plus fine. Ces éléments sont utiles la veille stratégique.

- Les couples entités-verbes Laurent Desbonnets - expliquer - Berlin ou Laurent Desbonnets - correspondre - Berlin permettent de comprendre le rôle de Laurent Desbonnets relativement à l'actualité allemande rapportée à France-info.

# Question 16:  Utilisation de l’arbre de dépendance syntaxique ?

L'utilisation d'un arbre de dépendance syntaxique permettrait d'affiner l'analyse des relations entre entités en étudiant la structure grammaticale des phrases. En identifiant les dépendances syntaxiques, on peut mieux isoler les verbes qui relient directement les entités en fonction de leur rôle grammatical (sujet, objet, complément). Cela permettrait de discriminer les verbes réellement significatifs pour une relation de ceux qui sont seulement liés à la structure de la phrase sans exprimer une interaction concrète entre les entités.

Par exemple dans le cas d'application 3 réalisé à la question 14 où l'on a

Phrase  : Emmanuel Macron est en effet attendu mercredi à Bruxelles avant de s'envoler pour Mayotte.
Verbes  : attendre, envoler

Mayotte est complément circonstantiel de but pour le verbe "envoler" dont Emmanuel Macro est le sujet, tandis que c'est Bruxelle qui est en relation avec "attendre" dont il est complément circonstantiel de lieu.